STAGING - 21.0.0.12 GA Blog (#1922)
OpenLiberty/blogs: posts/2021-11-26-jakarta-ee-9.1.adoc

---
layout: post
title: "Jakarta EE 9.1 support and configurable response headers in Open Liberty 21.0.0.12"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ryan-storey
author_github: https://github.com/ryan-storey
seo-title: Jakarta EE 9.1 support and configurable response headers in Open Liberty 21.0.0.12 - OpenLiberty.io
seo-description: Jakarta EE 9.1 support is now available as part of Open Liberty 21.0.0.12, alongside configurable response headers, which offer more granular control over response headers! Several significant bug fixes are also part of this release.
blog_description: Jakarta EE 9.1 support is now available as part of Open Liberty 21.0.0.12, alongside configurable response headers, which offer more granular control over response headers! Several significant bug fixes are also part of this release.

> open-graph-image: https://openliberty.io/img/twitter_card.jpg

Reviewer: switch to jakartaee logo png

 
In link:{url-about}[Open Liberty] 21.0.0.12:

* <<jee_9.1, Jakarta EE 9.1 support>>
* <<response_headers, Configurable Response Headers>>
* <<bugs, Notable bug fixes>>

Along with the new features and functions added to the runtime, we've also made <<guides, updates to our guides>>.

[#run]
== Run your apps using 21.0.0.12

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.12</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[21.0.0.12,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#jee_9.1]
== Jakarta EE 9.1 support

> image::img/blog/jakarta_ee_logo_schooner_color_horizontal_default.svg[align="center"]

Reviewer: change to png

Jakarta EE 9.1 support is now available in Open Liberty!!  This support allows you to run Jakarta EE 9.1 applications using Java 8, 11, or 17 with other Open Liberty value-add features that are updated to support Jakarta EE 9.1.  Many of you have followed our progress for delivering Jakarta EE 9.1 via our beta releases and corresponding blogs, and we'd like to thank those that provided feedback along the way.  

> If you're targetting a new application for Jakarta EE 9.1, make sure to use the `jakarta` namespace.  For existing applications that you'd like to move from Java EE (and its `javax` namespace) to Jakarta EE (and its `jakarta` namespace) we recommend trying the Eclipse Transformer, an open source project originally developed by members of the Open Liberty team and then contributed to the Eclipse Foundation.  

Reviewer: ```suggestion
If you're targetting a new application for Jakarta EE 9.1, make sure to use the `jakarta` namespace.  For existing applications that you'd like to move from Java EE (and its `javax` namespace) to Jakarta EE (and its `jakarta` namespace) we recommend trying the [Eclipse Transformer](https://github.com/eclipse/transformer), an open source project originally developed by members of the Open Liberty team and then contributed to the Eclipse Foundation.  
```

* The API package has changed from `javax.ejb` to `jakarta.ejb`
* The `@Schedule` annotation is now repeatable
* The following API methods have been removed:
** `javax.ejb.EJBContext.getCallerIdentity()` -> use `getCallerPrincipal()`
** `javax.ejb.EJBContext.getEnvironment()` -> use `JNDI` lookup in `java:comp/env`
** `javax.ejb.EJBContext.isCallerInRole(java.security.Identity)` -> use `isCallerInRole(String)`
** `javax.ejb.SessionContext.getMessageContext()` (removed with `JAX-RPC`)
   
All other capabilities of Enterprise Beans remain the same as the prior specification version (3.2).

Enable the Jakarta EE 9.1 features in your app’s `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9.1 convenience feature to enable all of the Jakarta EE 9.1 features at once:

[source,xml]
----
<featureManager>
    <feature>jakartaee-9.1</feature>
</featureManager>
----

Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9.1 Web Profile features at once:

[source,xml]
----
<featureManager>
    <feature>webProfile-9.1</feature>
</featureManager>
----

> |=======================
> |Jakarta EE 8 feature name & short name	|Jakarta EE 9.1 feature name & short name
> 
> |Enterprise JavaBeans (ejb-3.2) | Jakarta Enterprise Beans 4.0 (enterpriseBeans-4.0)
> |Enterprise JavaBeans Home Interfaces (ejbHome-3.2) |Jakarta Enterprise Beans Home Interfaces 4.0 (enterpriseBeansHome-4.0)
> |Enterprise JavaBeans Lite (ejbLite-3.2) |Jakarta Enterprise Beans Lite 4.0 (enterpriseBeansLite-4.0)
> |Enterprise JavaBeans Persistent Timers (ejbPersistentTimer-3.2) |Jakarta Enterprise Beans Persistent Timers 4.0 (enterpriseBeansPersistentTimer-4.0)
> |Enterprise JavaBeans Remote (ejbRemote-3.2) |Jakarta Enterprise Beans Remote 4.0 (enterpriseBeansRemote-4.0)
> |Expression Language (el-3.0) |Jakarta Expression Language 4.0 (expressionLanguage-4.0)
> |Java Authorization Contract for Containers (jacc-1.5) |Jakarta Authorization 2.0 (appAuthorization-2.0)
> |Java Authentication SPI for Containers (jaspic-1.1) |Jakarta Authentication 2.0 (appAuthentication-2.0)
> |Java EE Full Platform 8.0 (javaee-8.0) |Jakarta EE Platform 9.1 (jakartaee-9.1)
> |Java EE 8 Application Client (javaeeClient-8.0) |Jakarta EE Application Client 9.1 (jakartaeeClient-9.1)
> |JavaMail (javaMail-1.6) |Jakarta Mail 2.0 (mail-2.0)
> |Java Architecture for XML Binding (jaxb-2.2) |Jakarta XML Binding 3.0 (xmlBinding-3.0)
> |Java API for RESTful Web Services (jaxrs-2.1) |Jakarta RESTful Web Services 3.0 (restfulWS-3.0)
> |Java RESTful Services Client (jaxrsClient-2.1) |Jakarta RESTful Web Services Client 3.0 (restfulWSClient-3.0)
> |Java Web Services (jaxws-2.2) |Jakarta XML Web Services 3.0 (xmlWS-3.0)
> |Java Connector Architecture (jca-1.7) |Jakarta Connectors 2.0 (connectors-2.0)
> |Java Connector Architecture Security Inflow (jcaInboundSecurity-1.0) |Jakarta Connectors Inbound Security 2.0 (connectorsInboundSecurity-2.0)
> |Java Message Service (jms-2.0) |Jakarta Messaging 3.0 (messaging-3.0)
> |Java Persistence API (jpa-2.2) |Jakarta Persistence 3.0 (persistence-3.0)
> |Java Persistence API Container (jpaContainer-2.2) |Jakarta Persistence Container 3.0 (persistenceContainer-3.0)
> |JavaServer Faces (jsf-2.3) |Jakarta Server Faces 3.0 (faces-3.0)
> |JavaServer Faces Container (jsfContainer-2.3) |Jakarta Server Faces Container 3.0 (facesContainer-3.0)
> |JavaServer Pages (jsp-2.3) |Jakarta Server Pages 3.0 (pages-3.0)
> |JMS Client for Message Server (wasJmsClient-2.0) |Messaging Server Client 3.0 (messagingClient-3.0)
> |Message Server Security (wasJmsSecurity-1.0) |Messaging Server Security 3.0 (messagingSecurity-3.0)
> |Message Server (wasJmsServer-1.0) |Messaging Server 3.0 (messagingServer-3.0)
> |=======================

Reviewer: ```suggestion
Although many of the Jakarta EE 9.1 features have only received a version update, the majority have also had their name changed.  The following table lists the features for which both the short name and the version number are changed. To update one of these features for Jakarta EE 9.1, you must change both the feature short name and version number in your `server.xml` file.

.Table 1: Jakarta EE 9.1 feature updates, short name and version
|===
|Jakarta EE 9.1 feature name | Java EE/Jakarta EE 8 short name and version | Jakarta EE 9.1 short name and version

| Jakarta Enterprise Beans
| feature:ejb-3.2[]
| feature:enterpriseBeans-4.0[]

| Jakarta Enterprise Beans Home Interfaces
| feature:ejbHome-3.2[]
| feature:enterpriseBeansHome-4.0[]

| Jakarta Enterprise Beans Lite
| feature:ejbLite-3.2[]
| feature:enterpriseBeansLite-4.0[]

| Jakarta Enterprise Beans Persistent Timers
| feature:ejbPersistentTimer-3.2[]
| feature:enterpriseBeansPersistentTimer-4.0[]

| Jakarta Enterprise Beans Remote
| feature:ejbRemote-3.2[]
| feature:enterpriseBeansRemote-4.0[]

| Jakarta Expression Language
| feature:el-3.0[]
| feature:expressionLanguage-4.0[]

| Jakarta Authorization
| feature:jacc-1.5[]
| feature:appAuthorization-2.0[]

| Jakarta Authentication
| feature:jaspic-1.1[]
| feature:appAuthentication-2.0[]

| Jakarta EE Platform
| feature:javaee-8.0[]
| feature:jakartaee-9.1[]

| Jakarta EE Application Client
| feature:javaeeClient-8.0[]
| feature:jakartaeeClient-9.1[]

| Jakarta Mail
| feature:javaMail-1.6[]
| feature:mail-2.0[]

| Jakarta XML Binding
| feature:jaxb-2.2[]
| feature:xmlBinding-3.0[]

| Jakarta RESTful Web Services
| feature:jaxrs-2.1[]
| feature:restfulWS-3.0[]

| Jakarta RESTful Web Services Client
| feature:jaxrsClient-2.1[]
| feature:restfulWSClient-3.0[]

| Jakarta XML Web Services
| feature:jaxws-2.2[]
| feature:xmlWS-3.0[]

| Jakarta Connectors
| feature:jca-1.7[]
| feature:connectors-2.0[]

| Jakarta Connectors Inbound Security
| feature:jcaInboundSecurity-1.0[]
| feature:connectorsInboundSecurity-2.0[]

| Jakarta Messaging
| feature:jms-2.0[]
| feature:messaging-3.0[]

| Jakarta Persistence
| feature:jpa-2.2[]
| feature:persistence-3.0[]

| Jakarta Persistence Container
| feature:jpaContainer-2.2[]
| feature:persistenceContainer-3.0[]

| Jakarta Server Faces
| feature:jsf-2.3[]
| feature:faces-3.0[]

| Jakarta Server Faces Container
| feature:jsfContainer-2.3[]
| feature:facesContainer-3.0[]

| Jakarta Server Pages
| feature:jsp-2.3[]
| feature:pages-3.0[]

| Messaging Server Client
| feature:wasJmsClient-2.0[]
| feature:messagingClient-3.0[]

| Messaging Server Security
| feature:wasJmsSecurity-1.0[]
| feature:messagingSecurity-3.0[]

| Messaging Server
| feature:wasJmsServer-1.0[]
| feature:messagingServer-3.0[]
|===

For a full overview of what has changed, visit the [Jakarta EE 9.1 feature updates](https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.12/jakarta-ee9-feature-updates.html) page.
```

.Table 1: Jakarta EE 9.1 feature updates, short name and version
|===
|Jakarta EE 9.1 feature name | Java EE/Jakarta EE 8 short name and version | Jakarta EE 9.1 short name and version

| Jakarta Enterprise Beans
| ejb-3.2
| enterpriseBeans-4.0

| Jakarta Enterprise Beans Home Interfaces
| ejbHome-3.2
| enterpriseBeansHome-4.0

| Jakarta Enterprise Beans Lite
| ejbLite-3.2
| enterpriseBeansLite-4.0

| Jakarta Enterprise Beans Persistent Timers
| ejbPersistentTimer-3.2
| enterpriseBeansPersistentTimer-4.0

| Jakarta Enterprise Beans Remote
| ejbRemote-3.2
| enterpriseBeansRemote-4.0

| Jakarta Expression Language
| el-3.0
| expressionLanguage-4.0

| Jakarta Authorization
| jacc-1.5
| appAuthorization-2.0

| Jakarta Authentication
| jaspic-1.1
| appAuthentication-2.0

| Jakarta EE Platform
| javaee-8.0
| jakartaee-9.1

| Jakarta EE Application Client
| javaeeClient-8.0
| jakartaeeClient-9.1

| Jakarta Mail
| javaMail-1.6
| mail-2.0

| Jakarta XML Binding
| jaxb-2.2
| xmlBinding-3.0

| Jakarta RESTful Web Services

> | jaxrs-2.1
> | restfulWS-3.0
> 
> | Jakarta RESTful Web Services Client
> | jaxrsClient-2.1
> | restfulWSClient-3.0
> 
> | Jakarta XML Web Services
> | jaxws-2.2
> | xmlWS-3.0
> 
> | Jakarta Connectors
> | jca-1.7
> | connectors-2.0
> 
> | Jakarta Connectors Inbound Security
> | jcaInboundSecurity-1.0
> | connectorsInboundSecurity-2.0
> 
> | Jakarta Messaging
> | jms-2.0
> | messaging-3.0
> 
> | Jakarta Persistence
> | jpa-2.2
> | persistence-3.0
> 
> | Jakarta Persistence Container
> | jpaContainer-2.2
> | persistenceContainer-3.0
> 
> | Jakarta Server Faces
> | jsf-2.3
> | faces-3.0
> 
> | Jakarta Server Faces Container
> | jsfContainer-2.3
> | facesContainer-3.0
> 
> | Jakarta Server Pages
> | jsp-2.3
> | pages-3.0
> 
> | Messaging Server Client
> | wasJmsClient-2.0
> | messagingClient-3.0
> 
> | Messaging Server Security
> | wasJmsSecurity-1.0
> | messagingSecurity-3.0
> 
> | Messaging Server
> | wasJmsServer-1.0
> | messagingServer-3.0
> |===
> 
> For a full overview of what has changed, visit the link:{url-prefix}/docs/latest/jakarta-ee9-feature-updates.html[Jakarta EE 9.1 feature updates] page.
> 
> For more information, see the link:{url-prefix}/docs/latest/reference/jakarta-ee9-feature-updates.html[Jakarta EE 9.1 feature updates] documentation.
> For details regarding the APIs and SPIs, check out the link:{url-prefix}/docs/latest/reference/javadoc/liberty-jakartaee9.1-javadoc.html[Jakarta EE 9.1 javadoc].

Reviewer: ```suggestion
Although many of the Jakarta EE 9.1 features have only received a version update, the majority have also had their name changed.  The following table lists the features for which both the short name and the version number are changed. To update one of these features for Jakarta EE 9.1, you must change both the feature short name and version number in your `server.xml` file.

.Jakarta EE 9.1 feature updates, short name and version
|===
|Jakarta EE 9.1 feature name | Java EE/Jakarta EE 8 short name and version | Jakarta EE 9.1 short name and version

| Jakarta Enterprise Beans
| ejb-3.2
| enterpriseBeans-4.0

| Jakarta Enterprise Beans Home Interfaces
| ejbHome-3.2
| enterpriseBeansHome-4.0

| Jakarta Enterprise Beans Lite
| ejbLite-3.2
| enterpriseBeansLite-4.0

| Jakarta Enterprise Beans Persistent Timers
| ejbPersistentTimer-3.2
| enterpriseBeansPersistentTimer-4.0

| Jakarta Enterprise Beans Remote
| ejbRemote-3.2
| enterpriseBeansRemote-4.0

| Jakarta Expression Language
| el-3.0
| expressionLanguage-4.0

| Jakarta Authorization
| jacc-1.5
| appAuthorization-2.0

| Jakarta Authentication
| jaspic-1.1
| appAuthentication-2.0

| Jakarta EE Platform
| javaee-8.0
| jakartaee-9.1

| Jakarta EE Application Client
| javaeeClient-8.0
| jakartaeeClient-9.1

| Jakarta Mail
| javaMail-1.6
| mail-2.0

| Jakarta XML Binding
| jaxb-2.2
| xmlBinding-3.0

| Jakarta RESTful Web Services
| jaxrs-2.1
| restfulWS-3.0

| Jakarta RESTful Web Services Client
| jaxrsClient-2.1
| restfulWSClient-3.0

| Jakarta XML Web Services
| jaxws-2.2
| xmlWS-3.0

| Jakarta Connectors
| jca-1.7
| connectors-2.0

| Jakarta Connectors Inbound Security
| jcaInboundSecurity-1.0
| connectorsInboundSecurity-2.0

| Jakarta Messaging
| jms-2.0
| messaging-3.0

| Jakarta Persistence
| jpa-2.2
| persistence-3.0

| Jakarta Persistence Container
| jpaContainer-2.2
| persistenceContainer-3.0

| Jakarta Server Faces
| jsf-2.3
| faces-3.0

| Jakarta Server Faces Container
| jsfContainer-2.3
| facesContainer-3.0

| Jakarta Server Pages
| jsp-2.3
| pages-3.0

| Messaging Server Client
| wasJmsClient-2.0
| messagingClient-3.0

| Messaging Server Security
| wasJmsSecurity-1.0
| messagingSecurity-3.0

| Messaging Server
| wasJmsServer-1.0
| messagingServer-3.0
|===

For a full overview of what has changed, visit the link:{url-prefix}/docs/latest/jakarta-ee9-feature-updates.html[Jakarta EE 9.1 feature updates] page.

To enable the Jakarta EE 9.1 features, add the corresponding feature to your `server.xml`. You can enable either the individual features you want or you can add the Jakarta EE 9.1 convenience features.  For example, to enable all of the Jakarta EE 9.1 features at once add:

[source,xml]
----
<featureManager>
    <feature>jakartaee-9.1</feature>
</featureManager>
----

Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9.1 Web Profile features at once:

[source,xml]
----
<featureManager>
    <feature>webProfile-9.1</feature>
</featureManager>
----

For details regarding the APIs and SPIs, check out the link:{url-prefix}/docs/latest/reference/javadoc/liberty-jakartaee9.1-javadoc.html[Jakarta EE 9.1 javadoc].
```

* link:https://github.com/OpenLiberty/open-liberty/issues/19227[Ensure ServletRequestListener#requestDestroyed is always called]
+
We discovered a bug where the `ServletRequestListener#requestDestroyed` call does not occur, if an exception occurs during async servlet while an `appSecurity-x.0` is enabled. For this bug to occur, two conditions must be met: the `webContainer` property `deferServletRequestListenerDestroyOnError` is true and an `appSecurity-x.0` feature is enabled. This issue has now been resolved.

* link:https://github.com/OpenLiberty/open-liberty/issues/19197[ClassCastException in JSP relating to JDT internal classes]
+
Open Liberty 21.0.0.9 introduced a bug where the following error occured for certain class lookups in JSP:
+
[source]
----
Error 500: java.lang.ClassCastException: class org.eclipse.jdt.internal.compiler.lookup.PlainPackageBinding cannot be cast to class org.eclipse.jdt.internal.compiler.lookup.TypeBinding (org.eclipse.jdt.internal.compiler.lookup.PlainPackageBinding and org.eclipse.jdt.internal.compiler.lookup.TypeBinding are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @3522bc53)
----
+

> This issue has now been fixed, as the private `isPackage(String)` method needed to be updated to correctly identify that `org.apache.jsp.test_jsp$1InnerClass` is a class and not a package.

Reviewer 2: I think it's better to leave this out, and only say "This issue has now been fixed".

Reviewer 2: Actually, you can leave it. But change the example to `com.ibm._jsp._test_JSP$1InnerClass` instead.

Reviewer: ```suggestion
This issue has now been fixed.
```

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.12</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[21.0.0.12,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#jee_9.1]
== Jakarta EE 9.1 support

image::img/blog/jakarta_ee.png[align="center"]

Jakarta EE 9.1 support is now available in Open Liberty!!  This support allows you to run Jakarta EE 9.1 applications using Java 8, 11, or 17 with other Open Liberty value-add features that are updated to support Jakarta EE 9.1.  Many of you have followed our progress for delivering Jakarta EE 9.1 via our beta releases and corresponding blogs, and we'd like to thank those that provided feedback along the way.  

> If you're targetting a new application for Jakarta EE 9.1, make sure to use the `jakarta` namespace.  For existing applications that you'd like to move from Java EE (and its `javax` namespace) to Jakarta EE (and its `jakarta` namespace) we recommend trying the [Eclipse Transformer](https://github.com/eclipse/transformer), an open source project originally developed by members of the Open Liberty team and then contributed to the Eclipse Foundation.  

Reviewer: ```suggestion
If you're targeting a new application for Jakarta EE 9.1, make sure to use the `jakarta` namespace.  For existing applications that you'd like to move from Java EE (and its `javax` namespace) to Jakarta EE (and its `jakarta` namespace) we recommend trying the link:https://github.com/eclipse/transformer[Eclipse Transformer], an open source project originally developed by members of the Open Liberty team and then contributed to the Eclipse Foundation.  
```

Enable the Jakarta EE 9.1 features in your app’s `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9.1 convenience feature to enable all of the Jakarta EE 9.1 features at once:

[source,xml]
----
<featureManager>
    <feature>jakartaee-9.1</feature>
</featureManager>
----

Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9.1 Web Profile features at once:

[source,xml]
----
<featureManager>
    <feature>webProfile-9.1</feature>
</featureManager>
----

Although many of the Jakarta EE 9.1 features have only received a version update, the majority have also had their name changed.  The following table lists the features for which both the short name and the version number are changed. To update one of these features for Jakarta EE 9.1, you must change both the feature short name and version number in your `server.xml` file.

.Table 1: Jakarta EE 9.1 feature updates, short name and version
|===
|Jakarta EE 9.1 feature name | Java EE/Jakarta EE 8 short name and version | Jakarta EE 9.1 short name and version

| Jakarta Enterprise Beans

> | feature:ejb-3.2[]

Reviewer 3: What's with the feature:name[]?  It doesn't format at all when I am looking at the preview.  Seems wrong.   Am I missing something?

Reviewer: rushed copy and paste; fixing now

| feature:enterpriseBeans-4.0[]

| Jakarta Enterprise Beans Home Interfaces
| feature:ejbHome-3.2[]
| feature:enterpriseBeansHome-4.0[]

| Jakarta Enterprise Beans Lite
| feature:ejbLite-3.2[]
| feature:enterpriseBeansLite-4.0[]

| Jakarta Enterprise Beans Persistent Timers
| feature:ejbPersistentTimer-3.2[]
| feature:enterpriseBeansPersistentTimer-4.0[]

| Jakarta Enterprise Beans Remote
| feature:ejbRemote-3.2[]
| feature:enterpriseBeansRemote-4.0[]

| Jakarta Expression Language
| feature:el-3.0[]
| feature:expressionLanguage-4.0[]

| Jakarta Authorization
| feature:jacc-1.5[]
| feature:appAuthorization-2.0[]

| Jakarta Authentication
| feature:jaspic-1.1[]
| feature:appAuthentication-2.0[]

| Jakarta EE Platform
| feature:javaee-8.0[]
| feature:jakartaee-9.1[]

| Jakarta EE Application Client
| feature:javaeeClient-8.0[]
| feature:jakartaeeClient-9.1[]

| Jakarta Mail
| feature:javaMail-1.6[]
| feature:mail-2.0[]

> | Jakarta XML Binding
> | feature:jaxb-2.2[]
> | feature:xmlBinding-3.0[]
> 
> | Jakarta RESTful Web Services
> | feature:jaxrs-2.1[]
> | feature:restfulWS-3.0[]
> 
> | Jakarta RESTful Web Services Client
> | feature:jaxrsClient-2.1[]
> | feature:restfulWSClient-3.0[]
> 
> | Jakarta XML Web Services
> | feature:jaxws-2.2[]
> | feature:xmlWS-3.0[]
> 
> | Jakarta Connectors
> | feature:jca-1.7[]
> | feature:connectors-2.0[]
> 
> | Jakarta Connectors Inbound Security
> | feature:jcaInboundSecurity-1.0[]
> | feature:connectorsInboundSecurity-2.0[]
> 
> | Jakarta Messaging
> | feature:jms-2.0[]
> | feature:messaging-3.0[]
> 
> | Jakarta Persistence
> | feature:jpa-2.2[]
> | feature:persistence-3.0[]
> 
> | Jakarta Persistence Container
> | feature:jpaContainer-2.2[]
> | feature:persistenceContainer-3.0[]
> 
> | Jakarta Server Faces
> | feature:jsf-2.3[]
> | feature:faces-3.0[]
> 
> | Jakarta Server Faces Container
> | feature:jsfContainer-2.3[]
> | feature:facesContainer-3.0[]
> 
> | Jakarta Server Pages
> | feature:jsp-2.3[]
> | feature:pages-3.0[]
> 
> | Messaging Server Client
> | feature:wasJmsClient-2.0[]
> | feature:messagingClient-3.0[]
> 
> | Messaging Server Security
> | feature:wasJmsSecurity-1.0[]
> | feature:messagingSecurity-3.0[]
> 
> | Messaging Server
> | feature:wasJmsServer-1.0[]
> | feature:messagingServer-3.0[]
> |===

Reviewer: ```suggestion
.Table 1: Jakarta EE 9.1 feature updates, short name and version
|===
|Jakarta EE 9.1 feature name | Java EE/Jakarta EE 8 short name and version | Jakarta EE 9.1 short name and version

| Jakarta Enterprise Beans
| ejb-3.2
| enterpriseBeans-4.0

| Jakarta Enterprise Beans Home Interfaces
| ejbHome-3.2
| enterpriseBeansHome-4.0

| Jakarta Enterprise Beans Lite
| ejbLite-3.2
| enterpriseBeansLite-4.0

| Jakarta Enterprise Beans Persistent Timers
| ejbPersistentTimer-3.2
| enterpriseBeansPersistentTimer-4.0

| Jakarta Enterprise Beans Remote
| ejbRemote-3.2
| enterpriseBeansRemote-4.0

| Jakarta Expression Language
| el-3.0
| expressionLanguage-4.0

| Jakarta Authorization
| jacc-1.5
| appAuthorization-2.0

| Jakarta Authentication
| jaspic-1.1
| appAuthentication-2.0

| Jakarta EE Platform
| javaee-8.0
| jakartaee-9.1

| Jakarta EE Application Client
| javaeeClient-8.0
| jakartaeeClient-9.1

| Jakarta Mail
| javaMail-1.6
| mail-2.0

| Jakarta XML Binding
| jaxb-2.2
| xmlBinding-3.0

| Jakarta RESTful Web Services
| jaxrs-2.1
| restfulWS-3.0

| Jakarta RESTful Web Services Client
| jaxrsClient-2.1
| restfulWSClient-3.0

| Jakarta XML Web Services
| jaxws-2.2
| xmlWS-3.0

| Jakarta Connectors
| jca-1.7
| connectors-2.0

| Jakarta Connectors Inbound Security
| jcaInboundSecurity-1.0
| connectorsInboundSecurity-2.0

| Jakarta Messaging
| jms-2.0
| messaging-3.0

| Jakarta Persistence
| jpa-2.2
| persistence-3.0

| Jakarta Persistence Container
| jpaContainer-2.2
| persistenceContainer-3.0

| Jakarta Server Faces
| jsf-2.3
| faces-3.0

| Jakarta Server Faces Container
| jsfContainer-2.3
| facesContainer-3.0

| Jakarta Server Pages
| jsp-2.3
| pages-3.0

| Messaging Server Client
| wasJmsClient-2.0
| messagingClient-3.0

| Messaging Server Security
| wasJmsSecurity-1.0
| messagingSecurity-3.0

| Messaging Server
| wasJmsServer-1.0
| messagingServer-3.0
|===
```

> For a full overview of what has changed, visit the [Jakarta EE 9.1 feature updates](https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.12/jakarta-ee9-feature-updates.html) page.

Reviewer 3: ```suggestion
For a full overview of what has changed, visit the link:{url-prefix}/docs/latest/jakarta-ee9-feature-updates.html[Jakarta EE 9.1 feature updates] page.
```

+
When a websocket connection is started, it starts as an HTTP connection. Previously, If an error occured during the transition between an HTTP and a websocket connection, which was known to be a websocket upgrade, the error processing would neglect to decrement a connection counter, which then caused the server to believe there is an open connection during server stop.
There were two scenarios where these quiesce errors would occur:
+
** When a read error occured during the transition between an HTTP and a websocket connection, the error processing neglected to decrement a connection counter, which then causes the server to believe there is an open connection during server stop.
+
** If a client immediately closed the websocket connection after it was opened, the original upgrade request handling may not have had enough time to close properly on the server. Once again, the connection counter failed to decrement leading the server to believe there is an open connection during the server stop.
+
This issue has been fixed by adding a new flag called `decrementNeeded` which helps to ensure that the decrement is not neglected.

> * link:https://github.com/OpenLiberty/open-liberty/issues/19193[Stop allowing creation of HTTP/2 streams if server is closing]

Reviewer 4: I think we should omit this issue completely - a partial fix was put in. I haven't created a new issue that will completely resolve this issue, but we do have an internal [PE/ZE issue](https://github.ibm.com/WASL3/MDV/issues/79) open.

Reviewer: removed

> * link:https://github.com/OpenLiberty/open-liberty/issues/19193[Stop allowing creation of HTTP/2 streams if server is closing]
> +
> Quiesce is meant to allow chains to gracefully finish active request/responses during the configured time threshold. During quiesce, the accept selectors should no longer allow for new connections to be made. However, since `HTTP/2` uses existing connections, it was previously possible for new `HTTP/2` streams to be created during the quiesce period. Thus, we worked to disallow creation of new streams when the kernel signals that the framework is closing down, which should help diminish the cases where quiesce fails to finish due to `HTTP/2` connections. Now, during quiesce, connections should begin closing down. In `HTTP/1.1` this implies disabling keep-alive. In `HTTP/2`, sending a `GOAWAY` once quiesce begins.

Reviewer: ```suggestion
```
